README.md
### Summary

A simple plugin to automatically adjust textarea height to fit text
using `rows` attribute

### Demo

A demo example can be found at [kipruss.github.io/jquery-autoresize](http://kipruss.github.io/jquery-autoresize)

### Usage

Load the plugin and use jQuery to find the desired `textarea` elements on which to call the plugin method. Setting a parameter `rows` in html code
is appropriate to avoid a shake of the textarea while a page is loading.
This parameter might be equal to parameter `minRows` or 1 if parameter
`minRows` is not specified.

```html
<textarea class="js-textarea" rows="1"></textarea>
<textarea class="js-textarea-minrows" rows="3"></textarea>
<textarea class="js-textarea-maxrows" rows="1"></textarea>
<textarea class="js-textarea-minmaxrows" rows="2"></textarea>

<script src="//ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js"></script>
<script src="js/jquery.autoresize.min.js"></script>
<script>
    $(function() {
        $('.js-textarea').autoResize();
        $('.js-textarea-minrows').autoResize({minRows: 3});
        $('.js-textarea-maxrows').autoResize({maxRows: 3});
        $('.js-textarea-minmaxrows').autoResize({minRows: 2, maxRows: 3});
    });
</script>
```

### How to build

  1. Install or update `node` and `npm`

    - Clear NPM's cache: `sudo npm cache clean -f`
    - Install a little helper called 'n': `sudo npm install -g n`
    - Install latest stable _NodeJS_ version: `sudo n stable`. Alternatively pick a specific version and install like this: `sudo n 0.8.20`

  2. Install latest _gulp_ version

    ```bash
    npm install gulpjs/gulp#4.0
    ```

  3. Add string in your `.bashrc` or launch in console:

    ```bash
    export PATH="./node_modules/.bin:$PATH"
    ```

  4. Install all packages: `npm install`

  5. Run building system

    - for run examples page `gulp dev` or just for building: `gulp build`

  6. Run js test

    - run `gulp lint`

### License

Released under the [MIT License](http://www.opensource.org/licenses/mit-license.php)

### Resources

* https://github.com/thomasjo/jquery-autoresize
* https://github.com/kipruss/lar4ik

changelog.md
## Changelog

##### v0.0.2 - 2016-08-31

* added `maxRows` parameter

##### v0.0.1 - 2016-08-30

* initial commit

gulpfile.js
'use strict';

const gulp = require('gulp');
const del = require('del');
const newer = require('gulp-newer');
const browserSync = require('browser-sync').create();
const notify = require('gulp-notify');
const uglify = require('gulp-uglify');
const rename = require('gulp-rename');
const combiner = require('stream-combiner2').obj;
const eslint = require('gulp-eslint');

const src_dir = './src';
const build_dir = './build';
const dist_dir = './dist';

gulp.task('clean', function() {
    return del(build_dir, dist_dir);
});

gulp.task('assets', function() {
    return gulp.src(src_dir + '/assets/**/*.*', {since: gulp.lastRun('assets')})
        .pipe(newer(build_dir))
        .pipe(gulp.dest(build_dir));
});

gulp.task('js', function() {
    return combiner(
        gulp.src(src_dir + '/js/*.*'),
        uglify(),
        rename({
            suffix: '.min'
        }),
        gulp.dest(build_dir + '/js'),
        gulp.dest(dist_dir)
    ).on('error', notify.onError());
});

gulp.task('build', gulp.series(
    'clean',
    gulp.parallel('js', 'assets'))
);

gulp.task('watch', function() {
    gulp.watch(src_dir + '/js/*.*', gulp.series('js'));
    gulp.watch(src_dir + '/assets/**/*.*', gulp.series('assets'));
});

gulp.task('serve', function() {
    browserSync.init({
        server: build_dir,
        port: 1234
    });
    browserSync.watch(build_dir + '/**/*.*').on('change', browserSync.reload);
});

gulp.task('dev',
    gulp.series('build', gulp.parallel('watch', 'serve'))
);

gulp.task('lint', function() {
  return gulp.src('src/js/*.*')
      .pipe(eslint())
      .pipe(eslint.failAfterError());
});


src/js/jquery.autoresize.js
/*!
 * jQuery textarea autoResize plugin v0.0.1 - 2016-08-30
 * http://github.com/kipruss/jquery-autoresize
 * Author: Konstantin Baev for Magic Web Solutions ( http://magicwebsolutions.co.uk/ )
 * Licensed under the MIT license
 */

;(function() {
    'use strict';

    var $ = this.$,

        defaults = {
            resize: $.noop,
            'minRows': 1
        };

    $.fn.autoResize = function(options) {
        var settings = $.extend({}, defaults, options);

        this.filter('textarea').each(function() {
            var $textarea = $(this).css({
                    'overflow-y': 'hidden',
                    'height': 'auto',
                    'resize': 'none'
                });

            var textarea = $textarea.get(0);

            var setBaseScrollHeight = function () {
                var savedValue = textarea.value;
                var savedRows = textarea.rows;
                textarea.value = '';
                textarea.rows = 1;
                textarea.baseScrollHeight = textarea.scrollHeight;
                textarea.value = savedValue;
                textarea.rows = savedRows;
            };

            var init = function () {
                if (!textarea.scrollHeight) return false;
                if (textarea.baseScrollHeight) return true;
                setBaseScrollHeight();
                return true;
            };

            var setRows = function () {
                if (!init()) return false;
                var lineHeight = parseInt($textarea.css('line-height'));
                textarea.rows = settings.minRows;
                var rows = Math.ceil((textarea.scrollHeight - textarea.baseScrollHeight) / lineHeight) + 1;
                textarea.rows = Math.max(rows, settings.minRows);
            };

            if (!textarea.rows || textarea.rows < settings.minRows) {
                textarea.rows = settings.minRows;
            }
            $textarea.off('.resize');

            if (supportsInputEvent()) {
                $textarea.on('input.resize', setRows);
            } else if (supportsPropertyChangeEvent()) {
                $textarea.on('propertychange.resize', setRows);
            } else {
                $textarea.on('keypress.resize', setRows);
            }
            $textarea.one('focus', setRows);
            setRows();
        });

        return this;
    };

    function supportsInputEvent() {
        if ('oninput' in document.body) {
            return true;
        }

        document.body.setAttribute('oninput', 'return');

        var supports = typeof document.body.oninput === 'function';
        delete document.body.oninput;
        return supports;
    }

    function supportsPropertyChangeEvent() {
        return 'onpropertychange' in document.body;
    }

}).call(this);
